File: src/ios/Sources/AnsightCore/PairingSessionConnector.swift
import Foundation
import Network

public final class PairingSessionConnector: PairingSessionConnecting, @unchecked Sendable {
    private let datagramClient: any PairingDatagramClient
    private let wifiStatusProvider: @Sendable () -> PairingWifiPreflightStatus
    private let simulatorLocalHostAddressProvider: @Sendable () -> String?

    public convenience init() {
        self.init(
            datagramClient: NetworkPairingDatagramClient(),
            simulatorLocalHostAddressProvider: PairingSimulatorLocalHostAddress.resolve
        )
    }

    init(
        datagramClient: any PairingDatagramClient,
        wifiStatusProvider: @escaping @Sendable () -> PairingWifiPreflightStatus = PairingWifiPreflight.getStatus,
        simulatorLocalHostAddressProvider: @escaping @Sendable () -> String? = { nil }
    ) {
        self.datagramClient = datagramClient
        self.wifiStatusProvider = wifiStatusProvider
        self.simulatorLocalHostAddressProvider = simulatorLocalHostAddressProvider
    }

    var localHostAddress: String? {
        normalizedHostAddress(simulatorLocalHostAddressProvider())
    }

    func connect(
        document: ParsedPairingDocument,
        clientName: String,
        options: PairingConnectionOptions?
    ) async -> PairingConnectionAttempt {
        let simulatorLocalHostAddress = normalizedHostAddress(simulatorLocalHostAddressProvider())
        let hostAddressCandidates = PairingHostAddressCandidates.resolve(
            discoveryHint: document.discoveryHint,
            hostAddressOverride: options?.hostAddressOverride,
            simulatorLocalHostAddress: simulatorLocalHostAddress
        )
        guard !hostAddressCandidates.isEmpty else {
            return .failure(
                "The scanned Ansight QR code does not contain a reachable Studio address.",
                code: PairingFailureCodes.hostAddressRequired
            )
        }

        let discoveryPort = options?.discoveryPort
            ?? document.discoveryHint?.discoveryPort
            ?? document.config.host.discoveryPort
        guard (1...65_535).contains(discoveryPort) else {
            return .failure(
                "Studio discovery port must be between 1 and 65535.",
                code: PairingFailureCodes.hostAddressRequired
            )
        }

        let hostNetworkCheckMessage = Self.hostNetworkCheckMessage(discoveryHint: document.discoveryHint)
        let hasSimulatorLocalHostCandidate = Self.hasSimulatorLocalHostCandidate(
            hostAddressCandidates,
            simulatorLocalHostAddress: simulatorLocalHostAddress
        )
        let wifiStatus = hasSimulatorLocalHostCandidate ? PairingWifiPreflightStatus.connected : wifiStatusProvider()
        if wifiStatus == .notConnected {
            return .failure(
                "This device must be on the same Wi-Fi network as Ansight Studio. \(hostNetworkCheckMessage)",
                code: PairingFailureCodes.wifiRequired
            )
        }
        if wifiStatus == .cellular, options?.allowCellularConnections != true {
            return .failure(
                "Cellular Studio connections are disabled.",
                code: PairingFailureCodes.wifiRequired
            )
        }

        let deviceId = PairingDeviceIdentity.resolve()
        let enrollmentAppId: String
        if document.config.appId.trimmingCharacters(in: .whitespacesAndNewlines) == PairingConfig.anyAppId {
            enrollmentAppId = Bundle.main.bundleIdentifier?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        } else {
            enrollmentAppId = document.config.appId
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }
        guard !enrollmentAppId.isEmpty, enrollmentAppId != PairingConfig.anyAppId else {
            return .failure(
                "Ansight could not resolve this app's bundle id for generic enrollment.",
                code: PairingFailureCodes.enrollmentRequired
            )
        }

        let isLocalEnrollment = document.config.configId.hasPrefix(
            PairingEnrollmentModes.localConfigPrefix
        )
        let discoveryTimeoutSeconds: TimeInterval = isLocalEnrollment ? 1 : 5
        var lastFailure: PairingConnectionAttempt?
        for hostAddress in hostAddressCandidates {
            let requestId = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
            let request = ConnectRequest(
                requestId: requestId,
                enrollmentMode: isLocalEnrollment
                    ? PairingEnrollmentModes.local
                    : PairingEnrollmentModes.invite,
                inviteId: document.config.configId,
                appId: enrollmentAppId,
                deviceId: deviceId,
                deviceName: clientName,
                accessToken: document.config.enrollment.accessToken
            )
            let requestData: Data
            do {
                requestData = try JSONEncoder.ansightEncoder.encode(request)
            } catch {
                return .failure(
                    "Failed to encode enrollment request: \(error.localizedDescription)",
                    code: PairingFailureCodes.udpBootstrapFailed
                )
            }

            let responseData: Data?
            do {
                responseData = try await datagramClient.sendConnectRequest(
                    requestData,
                    host: hostAddress,
                    port: discoveryPort,
                    timeoutSeconds: discoveryTimeoutSeconds
                )
            } catch {
                lastFailure = .failure(
                    "UDP enrollment failed for \(hostAddress): \(error.localizedDescription)",
                    code: PairingFailureCodes.udpBootstrapFailed
                )
                continue
            }
            guard let responseData else {
                lastFailure = .failure(
                    "No response from Studio at \(hostAddress). \(hostNetworkCheckMessage) Scan a fresh QR code if Studio's address changed.",
                    code: PairingFailureCodes.udpBootstrapTimeout
                )
                continue
            }

            let response: ConnectResponse
            do {
                response = try JSONDecoder.ansightDecoder.decode(ConnectResponse.self, from: responseData)
            } catch {
                return .failure(
                    "Studio enrollment response was malformed: \(error.localizedDescription)",
                    code: PairingFailureCodes.udpBootstrapFailed
                )
            }
            guard response.type == "ENROLLMENT_RESULT",
                  response.ver == 2,
                  response.requestId == requestId
            else {
                return .failure(
                    "Studio returned an unexpected enrollment response.",
                    code: PairingFailureCodes.udpBootstrapFailed
                )
            }
            guard response.accepted else {
                return .rejected(hostAddress: hostAddress, response: response)
            }
            guard let webSocketPort = response.webSocketPort,
                  let webSocketPath = response.webSocketPath,
                  let webSocketToken = response.webSocketToken,
                  !webSocketPath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
                  !webSocketToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            else {
                return .failure(
                    "Studio did not provide a WebSocket handoff.",
                    code: PairingFailureCodes.webSocketHandoffUnavailable
                )
            }

            var components = URLComponents()
            components.scheme = "ws"
            components.host = hostAddress
            components.port = webSocketPort
            components.path = webSocketPath.hasPrefix("/") ? webSocketPath : "/\(webSocketPath)"
            components.queryItems = [URLQueryItem(name: "token", value: webSocketToken)]
            guard let url = components.url else {
                return .failure(
                    "Studio WebSocket handoff was not a valid URL.",
                    code: PairingFailureCodes.webSocketHandoffUnavailable
                )
            }

            return .success(hostAddress: hostAddress, response: response, webSocketURL: url)
        }

        return lastFailure ?? .failure(
            "The scanned Ansight QR code does not contain a reachable Studio address.",
            code: PairingFailureCodes.hostAddressRequired
        )
    }

    private func normalizedHostAddress(_ address: String?) -> String? {
        guard let candidate = address?.trimmingCharacters(in: .whitespacesAndNewlines),
              !candidate.isEmpty
        else {
            return nil
        }
        return candidate
    }

    private static func hasSimulatorLocalHostCandidate(
        _ candidates: [String],
        simulatorLocalHostAddress: String?
    ) -> Bool {
        guard let simulatorLocalHostAddress else {
            return false
        }
        return candidates.contains {
            $0.caseInsensitiveCompare(simulatorLocalHostAddress) == .orderedSame
        }
    }

    private static func hostNetworkCheckMessage(discoveryHint: PairingDiscoveryHint?) -> String {
        if let wifiName = discoveryHint?.wifiName?.trimmingCharacters(in: .whitespacesAndNewlines),
           !wifiName.isEmpty {
            return "Last known Studio Wi-Fi: \(wifiName)."
        }
        return "Check that both devices are on the same Wi-Fi network."
    }
}

enum PairingDeviceIdentity {
    private static let key = "ai.ansight.enrollment.device-id"
    private static let accessTokenKey = "ai.ansight.enrollment.local-access-token"
    private static let lock = NSLock()

    static func resolve() -> String {
        lock.lock()
        defer { lock.unlock() }

        if let existing = UserDefaults.standard.string(forKey: key)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !existing.isEmpty {
            return existing
        }

        let deviceId = "apple.\(UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased())"
        UserDefaults.standard.set(deviceId, forKey: key)
        return deviceId
    }

    static func resolveAccessToken() -> String {
        lock.lock()
        defer { lock.unlock() }

        if let existing = UserDefaults.standard.string(forKey: accessTokenKey)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !existing.isEmpty {
            return existing
        }

        let bytes = Data((0..<32).map { _ in UInt8.random(in: .min ... .max) })
        let accessToken = bytes.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
        UserDefaults.standard.set(accessToken, forKey: accessTokenKey)
        return accessToken
    }
}

File: src/dotnet/Ansight.Core/Pairing/PairingSessionConnector.cs
using System.Net;
using System.Net.Sockets;
using System.Net.WebSockets;
using System.Text.Json;

namespace Ansight.Pairing;

internal sealed class PairingSessionConnector
{
    private readonly Func<PairingWifiPreflightStatus> wifiStatusProvider;
    private readonly Func<string?> simulatorLocalHostAddressProvider;

    public PairingSessionConnector()
        : this(PairingWifiPreflight.GetStatus, PairingSimulatorLocalHostAddress.Resolve)
    {
    }

    internal PairingSessionConnector(Func<PairingWifiPreflightStatus> wifiStatusProvider)
        : this(wifiStatusProvider, PairingSimulatorLocalHostAddress.Resolve)
    {
    }

    internal PairingSessionConnector(
        Func<PairingWifiPreflightStatus> wifiStatusProvider,
        Func<string?> simulatorLocalHostAddressProvider)
    {
        this.wifiStatusProvider = wifiStatusProvider ?? throw new ArgumentNullException(nameof(wifiStatusProvider));
        this.simulatorLocalHostAddressProvider = simulatorLocalHostAddressProvider
                                                 ?? throw new ArgumentNullException(nameof(simulatorLocalHostAddressProvider));
    }

    public async Task<PairingConnectionAttempt> ConnectAsync(
        ParsedPairingDocument document,
        string clientName,
        PairingConnectionOptions? options,
        IProgress<HostConnectionProgressUpdate>? progress,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(document);

        var simulatorLocalHostAddress = ResolveSimulatorLocalHostAddress();
        var config = document.Config;
        var hostAddressCandidates = PairingDiscoveryHintHostAddresses.ResolveCandidates(
            document.DiscoveryHint,
            options?.HostAddressOverride,
            simulatorLocalHostAddress);
        if (hostAddressCandidates.Length == 0)
        {
            return PairingConnectionAttempt.FromFailure(
                "A current host address is required. Import a fresh pairing config or compact pairing config code.",
                PairingFailureCodes.HostAddressRequired);
        }

        var hostNetworkCheckMessage = BuildHostNetworkCheckMessage(document.DiscoveryHint);
        if (!HasSimulatorLocalHostCandidate(hostAddressCandidates, simulatorLocalHostAddress) &&
            wifiStatusProvider() == PairingWifiPreflightStatus.NotConnected)
        {
            var wifiRequiredMessage =
                $"Ansight is unavailable because this device is not connected to Wi-Fi. {hostNetworkCheckMessage}";

            HostPairingProgressReporter.Report(
                progress,
                HostConnectionProgressKind.Connection,
                wifiRequiredMessage,
                source: HostConnectionSource.HostConnection,
                reasonCode: PairingFailureCodes.WifiRequired);

            return PairingConnectionAttempt.FromFailure(
                wifiRequiredMessage,
                PairingFailureCodes.WifiRequired);
        }

        var discoveryPort = PairingDiscoveryPortResolver.Resolve(document, options?.DiscoveryPort);
        var usingHostOverride = !string.IsNullOrWhiteSpace(options?.HostAddressOverride);
        HostPairingProgressReporter.Report(
            progress,
            HostConnectionProgressKind.Connection,
            usingHostOverride
                ? $"Using host override address: {hostAddressCandidates[0]}"
                : $"Using pairing config host address candidates: {string.Join(", ", hostAddressCandidates)}",
            source: HostConnectionSource.HostConnection);

        PairingConnectionAttempt? lastFailure = null;
        for (var index = 0; index < hostAddressCandidates.Length; index++)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var hostAddressCandidate = hostAddressCandidates[index];
            if (!TryResolveHostAddress(hostAddressCandidate, out var hostAddress))
            {
                var invalidAddressMessage = $"Pairing host address '{hostAddressCandidate}' is not a valid IP address.";
                HostPairingProgressReporter.Report(
                    progress,
                    HostConnectionProgressKind.Connection,
                    invalidAddressMessage,
                    source: HostConnectionSource.HostConnection,
                    reasonCode: PairingFailureCodes.HostAddressRequired);

                lastFailure = PairingConnectionAttempt.FromFailure(
                    invalidAddressMessage,
                    PairingFailureCodes.HostAddressRequired);
                continue;
            }

            HostPairingProgressReporter.Report(
                progress,
                HostConnectionProgressKind.Connection,
                hostAddressCandidates.Length == 1
                    ? $"Connecting to host at {hostAddress}:{discoveryPort}"
                    : $"Connecting to host at {hostAddress}:{discoveryPort} ({index + 1}/{hostAddressCandidates.Length})",
                source: HostConnectionSource.HostConnection);

            using var connectTimeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            connectTimeout.CancelAfter(TimeSpan.FromSeconds(5));

            ConnectResponse? connectResponse;
            try
            {
                connectResponse = await SendConnectRequestAsync(
                    config,
                    clientName,
                    hostAddress!,
                    discoveryPort,
                    connectTimeout.Token);
            }
            catch (SocketException ex)
            {
                lastFailure = PairingConnectionAttempt.FromFailure(
                    $"UDP connect failed for {hostAddress}: {ex.Message}",
                    PairingFailureCodes.UdpBootstrapFailed);
                continue;
            }

            if (connectResponse is null)
            {
                lastFailure = PairingConnectionAttempt.FromFailure(
                    $"No connect response from host at {hostAddress}. {hostNetworkCheckMessage} The remembered host address may be stale. Import a fresh pairing QR code or enter the host IP manually.",
                    PairingFailureCodes.UdpBootstrapTimeout);
                continue;
            }

            HostPairingProgressReporter.Report(
                progress,
                HostConnectionProgressKind.Connection,
                $"Host response: {connectResponse.Message}",
                source: HostConnectionSource.HostConnection,
                reasonCode: connectResponse.Reason);
            if (!string.IsNullOrWhiteSpace(connectResponse.ReasonMessage))
            {
                HostPairingProgressReporter.Report(
                    progress,
                    HostConnectionProgressKind.Connection,
                    $"Reason: {connectResponse.ReasonMessage}",
                    source: HostConnectionSource.HostConnection,
                    reasonCode: connectResponse.Reason);
            }

            HostPairingProgressReporter.Report(
                progress,
                HostConnectionProgressKind.Connection,
                $"Reason code: {connectResponse.Reason}",
                source: HostConnectionSource.HostConnection,
                reasonCode: connectResponse.Reason);
            HostPairingProgressReporter.Report(
                progress,
                HostConnectionProgressKind.Connection,
                $"Accepted: {connectResponse.Accepted}",
                source: HostConnectionSource.HostConnection,
                reasonCode: connectResponse.Reason);

            if (!connectResponse.Accepted)
            {
                return PairingConnectionAttempt.FromRejected(hostAddress!, connectResponse);
            }

            if (connectResponse.WebSocketPort is null ||
                string.IsNullOrWhiteSpace(connectResponse.WebSocketPath) ||
                string.IsNullOrWhiteSpace(connectResponse.WebSocketToken))
            {
                return PairingConnectionAttempt.FromFailure(
                    "Host did not provide a WebSocket handoff.",
                    PairingFailureCodes.WebSocketHandoffUnavailable);
            }

            var wsUri = BuildWebSocketUri(
                hostAddress!,
                connectResponse.WebSocketPort.Value,
                connectResponse.WebSocketPath,
                connectResponse.WebSocketToken);
            HostPairingProgressReporter.Report(
                progress,
                HostConnectionProgressKind.Connection,
                $"Opening WebSocket: {wsUri}",
                source: HostConnectionSource.Transport);

            var connectedSocket = await ConnectWebSocketWithRetryAsync(wsUri, cancellationToken);
            if (connectedSocket is null)
            {
                return PairingConnectionAttempt.FromFailure(
                    "WebSocket endpoint did not become reachable in time.",
                    PairingFailureCodes.WebSocketEndpointUnreachable);
            }

            return PairingConnectionAttempt.FromSuccess(hostAddress!, connectResponse, connectedSocket);
        }

        return lastFailure ?? PairingConnectionAttempt.FromFailure(
            "A current host address is required. Import a fresh pairing config or compact pairing config code.",
            PairingFailureCodes.HostAddressRequired);
    }

    private static async Task<ClientWebSocket?> ConnectWebSocketWithRetryAsync(Uri wsUri, CancellationToken cancellationToken)
    {
        const int maxAttempts = 12;

        for (var attempt = 1; attempt <= maxAttempts; attempt++)
        {
            var webSocket = new ClientWebSocket();

            try
            {
                using var connectTimeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                connectTimeout.CancelAfter(TimeSpan.FromSeconds(2));
                await webSocket.ConnectAsync(wsUri, connectTimeout.Token);

                if (webSocket.State == WebSocketState.Open)
                {
                    return webSocket;
                }
            }
            catch when (attempt < maxAttempts)
            {
                // Retry.
            }
            catch
            {
                webSocket.Dispose();
                throw;
            }

            webSocket.Dispose();

            if (attempt < maxAttempts)
            {
                await Task.Delay(250, cancellationToken);
            }
        }

        return null;
    }

    private static async Task<ConnectResponse?> SendConnectRequestAsync(
        PairingConfig config,
        string clientName,
        IPAddress hostAddress,
        int discoveryPort,
        CancellationToken cancellationToken)
    {
        using var udpClient = new UdpClient(hostAddress.AddressFamily);

        var request = new ConnectRequest
        {
            Type = "CONNECT_REQ",
            Ver = 1,
            ConfigId = config.ConfigId,
            OneTimeToken = config.OneTimeToken,
            AppId = config.AppId,
            ClientName = clientName,
            ProcessSessionId = ProcessSessionIdentity.Current
        };

        var bytes = JsonSerializer.SerializeToUtf8Bytes(request, PairingJson.Compact);
        await udpClient.SendAsync(bytes, bytes.Length, new IPEndPoint(hostAddress, discoveryPort));

        while (!cancellationToken.IsCancellationRequested)
        {
            UdpReceiveResult receiveResult;
            try
            {
                receiveResult = await udpClient.ReceiveAsync(cancellationToken);
            }
            catch (OperationCanceledException)
            {
                return null;
            }

            if (!Equals(receiveResult.RemoteEndPoint.Address, hostAddress))
            {
                continue;
            }

            try
            {
                var response = JsonSerializer.Deserialize<ConnectResponse>(receiveResult.Buffer, PairingJson.Compact);
                if (response is not null && string.Equals(response.Type, "CONNECT_RESP", StringComparison.Ordinal))
                {
                    return response;
                }
            }
            catch
            {
                // Ignore malformed packets.
            }
        }

        return null;
    }

    private static Uri BuildWebSocketUri(IPAddress hostAddress, int port, string path, string token)
    {
        var builder = new UriBuilder(Uri.UriSchemeWs, hostAddress.ToString(), port)
        {
            Path = path,
            Query = $"token={Uri.EscapeDataString(token)}"
        };
        return builder.Uri;
    }

    private static bool TryResolveHostAddress(string? hostAddressText, out IPAddress? hostAddress)
    {
        hostAddress = null;
        if (string.IsNullOrWhiteSpace(hostAddressText))
        {
            return false;
        }

        return IPAddress.TryParse(hostAddressText.Trim(), out hostAddress);
    }

    private string? ResolveSimulatorLocalHostAddress()
    {
        try
        {
            var address = simulatorLocalHostAddressProvider();
            return string.IsNullOrWhiteSpace(address) ? null : address.Trim();
        }
        catch (Exception ex)
        {
            Logger.Info($"Simulator host-address detection failed: {ex.Message}");
            return null;
        }
    }

    private static bool HasSimulatorLocalHostCandidate(string[] candidates, string? simulatorLocalHostAddress)
    {
        if (string.IsNullOrWhiteSpace(simulatorLocalHostAddress))
        {
            return false;
        }

        return candidates.Contains(simulatorLocalHostAddress.Trim(), StringComparer.OrdinalIgnoreCase);
    }

    internal static string BuildHostNetworkCheckMessage(PairingDiscoveryHint? discoveryHint)
    {
        var wifiName = discoveryHint?.WifiName?.Trim();
        if (!string.IsNullOrWhiteSpace(wifiName))
        {
            return $"Check that this device is on the same Wi-Fi network as the Ansight host. Last known host Wi-Fi: {wifiName}.";
        }

        return "Check that this device is on the same Wi-Fi network as the Ansight host.";
    }
}

internal sealed record PairingConnectionAttempt(
    bool Success,
    bool Accepted,
    string Message,
    IPAddress? HostAddress,
    ConnectResponse? ConnectResponse,
    ClientWebSocket? WebSocket,
    string? FailureCode)
{
    public static PairingConnectionAttempt FromFailure(string message, string? failureCode = null)
        => new(false, false, message, null, null, null, failureCode);

    public static PairingConnectionAttempt FromRejected(IPAddress hostAddress, ConnectResponse connectResponse)
        => new(false, false, connectResponse.ReasonMessage ?? connectResponse.Message, hostAddress, connectResponse, null, null);

    public static PairingConnectionAttempt FromSuccess(
        IPAddress hostAddress,
        ConnectResponse connectResponse,
        ClientWebSocket webSocket)
        => new(true, true, "Connected to host and WebSocket session is ready.", hostAddress, connectResponse, webSocket, null);
}
